src/v/raft/log_eviction_stm.cc

```diff
@@ -70,61 +210,135 @@ void log_eviction_stm::process_log_eviction() {
+        }
+
         /// Evict as much as data as possible ...
         eviction_offset = max_collectible_offset;
     }
 
+    auto segment_boundary_offset = eviction_offset;
+    if (event.event_type == eviction_event_type::kafka) {
+        /// In the event this method was triggered from a client request its
+        /// very likely that the requested eviction offset may fall between
+        /// segment and or batch boundaries. Locate the offset of the earliest
+        /// segment boundary to perform the truncation at.
+        const auto opt_segment_boundary_offset
+          = _raft->log().first_segment_boundary_after_offset(eviction_offset);
+        vassert(
+          opt_segment_boundary_offset.has_value(),
+          "Failed to find earliest offset in segment for eviction_offset: {}",
+          eviction_offset);
+
+        segment_boundary_offset = model::prev_offset(
+          *opt_segment_boundary_offset);
+        vassert(
+          segment_boundary_offset <= eviction_offset,
+          "first_segment_boundary_after_offset() must return an offset <= the "
+          "input");
+    }
+
     return _raft->visible_offset_monitor()
-      .wait(eviction_offset, model::no_timeout, _as)
-      .then([this, eviction_offset]() mutable {
+      .wait(segment_boundary_offset, model::no_timeout, _as)
+      .then([this, segment_boundary_offset, eviction_offset]() mutable {
           auto f = ss::now();
 
           if (_stm_manager) {
-              f = _raft->refresh_commit_index().then([this, eviction_offset] {
-                  return _stm_manager->ensure_snapshot_exists(eviction_offset);
-              });
+              f = _raft->refresh_commit_index().then(
+                [this, segment_boundary_offset] {
+                    return _stm_manager->ensure_snapshot_exists(
+                      segment_boundary_offset);
+                });
           }
 
-          return f.then([this, eviction_offset]() {
-              return _raft->write_snapshot(
-                write_snapshot_cfg(eviction_offset, iobuf()));
+          return f.then([this, segment_boundary_offset, eviction_offset]() {
+              /// Use next_offset so that delta calculation works even in the
+              /// case segment_boundary_offset is == model::offset::min
+              int64_t delta = model::next_offset(eviction_offset)
+                              - model::next_offset(segment_boundary_offset);
+              vassert(
+                delta >= 0,
+                "eviction_offset {} expected to always be >= "
+                "segment_boundary_offset {}",
+                eviction_offset,
+                segment_boundary_offset);
+              return _raft
+                ->write_snapshot(write_snapshot_cfg(
+                  segment_boundary_offset,
+                  log_start_offset_delta(delta),
+                  iobuf()))
+                .then([eviction_offset] { return eviction_offset; });
```

Comment:
What if we wouldn't store the `log_start_offset_delta` or `effective_start_offset` in `snapshot_metadata` but make it part of `iobuf()` this way we wouldn't have to change raft interface at all. the only part that would change is `cluster:partition` it would query the `_log_eviction_stm` for start offset i think that solution would be much cleaner as it wouldn't have to introduce delete offsets specific changes in general Raft algorithm implementation